File: source/downloads/code/LeetCode/Java/Find-Peak-Element.java
public class Solution {
    public int findPeakElement(int[] nums) {
        int n = nums.length;
        return bs(nums,0,n-1);   
    }

    public int bs(int[] nums, int start, int end){
        if(start==end)
            return start;
        int mid1 = start+(end-start)/2;
        int mid2 = mid1+1;
        if(nums[mid1]>nums[mid2])
            return bs(nums, start,mid1);
        else
            return bs(nums, mid2,end);
    }
}

public class Solution {
    public int findPeakElement(int[] nums){
        int end = nums.length-1;
        int start = 0;
        while(start<end){
            int mid1 = start+(end-start)/2;
            int mid2 = mid1+1;
            if(nums[mid1]>nums[mid2])
                end = mid1;
            else
                start = mid2;
        
        }
        return start;
    }

}

File: downloads/code/LeetCode/Python/Find-Peak-Element.py
class Solution(object):
    def findPeakElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 1:
            return 0

        start = 0
        end = len(nums)-1
        while start <= end:
            mid = (start+end)/2
            if mid-1 < 0 and mid+1 < len(nums):
                if nums[mid] > nums[mid+1]:
                    return mid
                else:
                    start = mid+1
            if mid+1 >= len(nums) and mid-1>=0:
                if nums[mid] > nums[mid-1]:
                    return mid
                else: 
                    end = mid-1

            if mid-1 >=0 and mid+1<len(nums):
                if nums[mid] > nums[mid-1] and nums[mid] > nums[mid+1]:
                    return mid
                elif nums[mid] < nums[mid-1]:
                    end = mid-1
                else: 
                    #nums[mid] < nums[mid+1]:
                    start = mid+1
        return start

if __name__ == "__main__":
    sol = Solution()
    sol.findPeakElement([3,2,1])
